hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+//!
+//! // Set Region Mismatch Interrupt
+//! icm_region2.set_rdm_int();
+//!
+//! // Setup region 3 to monitor memory
+//! icm_region_desc
+//!     .region3
+//!     .set_raddr(&message_region3_sha256);
+//! icm_region_desc.region3.rcfg.reset_rcfg_to_default();
+//! icm_region_desc.region3.rcfg.set_algo(icm_algorithm::SHA256);
+//! // Activate Compare Digest (should be true when comparing memory)
+//! icm_region_desc.region3.rcfg.set_cdwbn(true);
+//! // Digest Mismatch Interrupt Disable (enabled)
+//! icm_region_desc.region3.rcfg.set_dmien(false);
+//! // Wrap
+//! icm_region_desc.region3.rcfg.set_wrap(true);
+//!
+//! // Set Region Mismatch Interrupt
+//! icm_region3.set_rdm_int();
+//!
+//! // Modify regions to trigger interrupts
+//! message_region0_sha1[3] = 0xDEAD_BEEF;
+//! message_region1_sha1[4] = 0xDEAD_BEEF;
+//! message_region2_sha224[5] = 0xDEAD_BEEF;
+//! message_region3_sha256[6] = 0xDEAD_BEEF;
+//!
+//! icm.enable()
+use crate::pac::generic::Variant;
+use crate::pac::icm::uasr::URAT_A;
+
+use crate::pac::icm::*;
+use crate::typelevel::Sealed;
+use core::marker::PhantomData;
+
+/// Reexport the User SHA Algorithm
+pub use crate::icm::cfg::UALGO_A as icm_algorithm;
+
+/// IcmToken represents the peripheral and it encapsulates the hardware register
+struct IcmToken {
+    icm: crate::pac::ICM,
+}
+
+impl IcmToken {
+    /// Create a new instance of [`IcmToken`]
+    fn new(icm: crate::pac::ICM) -> Self {
+        Self { icm }
+    }
+
+    /// Integrity Check Module
+    #[inline]
+    fn icm(&self) -> &RegisterBlock {
+        &self.icm
+    }
+
+    /// Configuration
+    #[inline]
+    fn cfg(&self) -> &CFG {
+        &self.icm().cfg
+    }
+
+    /// Control
+    #[inline]
+    fn ctrl(&self) -> &CTRL {
+        &self.icm().ctrl
+    }
+
+    /// Region Descriptor Area Start Address
+    #[inline]
+    fn dscr(&self) -> &DSCR {
+        &self.icm().dscr
+    }
+
+    /// Region Hash Area Start Address
+    #[inline]
+    fn hash(&self) -> &HASH {
+        &self.icm().hash
+    }
+
+    /// Interrupt Disable
+    #[inline]
+    fn idr(&self) -> &IDR {
+        &self.icm().idr
+    }
+
+    /// Interrupt Enable
+    #[inline]
+    fn ier(&self) -> &IER {
+        &self.icm().ier
+    }
+
+    /// Interrupt Mask
+    #[inline]
+    pub fn imr(&self) -> &IMR {
+        &self.icm().imr
+    }
+
+    /// Interrupt Status
+    #[inline]
+    fn isr(&self) -> &ISR {
+        &self.icm().isr
+    }
+
+    /// Status
+    #[inline]
+    fn sr(&self) -> &SR {
+        &self.icm().sr
+    }
+
+    /// Undefined Access Status
+    #[inline]
+    fn uasr(&self) -> &UASR {
+        &self.icm().uasr
+    }
+}
+
+/// Struct useful for returning the interrupt status
+/// of the ICM. Provides methods for easy parsing of
+/// all the regions or via the `bitmask` argument
+/// narrow it down to the specific set of [`IcmRegionNum`]
+/// of interest.
+pub struct IcmInterrupt {
+    interrupt_vector: u32,
+}
```

Comment:
You might consider `bitflags` or some other crate for this type. I think it would end up being simpler for you and more useful for users. @jbeaurivage and I found it useful for a similar purpose in `sercom::v2`.

Reply:
Great idea, adopted!